tests/functional/test_discovery.py

```diff
@@ -0,0 +1,371 @@
+
+    def test_endpoint_discovery_required_with_env_var_default(self):
+        environ = os.environ.copy()
+        with mock.patch('os.environ', environ):
+            discovered_endpoint = 'https://discovered.domain'
+            client, http_stubber = self.create_client(
+                service_name="test-discovery-endpoint"
+            )
+            with http_stubber as stubber:
+                self.add_describe_endpoints_response(stubber, discovered_endpoint)
+                client.test_discovery_required(Foo="bar")
+                self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_required_with_env_var_default(self):
+        environ = os.environ.copy()
+        environ['AWS_ENDPOINT_DISCOVERY_ENABLED'] = 'auto'
+        with mock.patch('os.environ', environ):
+            discovered_endpoint = 'https://discovered.domain'
+            client, http_stubber = self.create_client(
+                service_name="test-discovery-endpoint", config=None
+            )
+            with http_stubber as stubber:
+                self.add_describe_endpoints_response(stubber, discovered_endpoint)
+                client.test_discovery_required(Foo="bar")
+                self.assertEqual(
+                    os.environ.get('AWS_ENDPOINT_DISCOVERY_ENABLED'), 'auto'
+                )
+                self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_optional_with_env_var_default(self):
+        self.environ['AWS_ENDPOINT_DISCOVERY_ENABLED'] = 'auto'
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=None
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(
+                os.environ.get('AWS_ENDPOINT_DISCOVERY_ENABLED'), 'auto'
+            )
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryOptional"
+            )
+
+    def test_endpoint_discovery_optional_with_env_var_enabled(self):
+        self.environ['AWS_ENDPOINT_DISCOVERY_ENABLED'] = "True"
+        discovered_endpoint = 'https://discovered.domain'
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint"
+        )
+        with http_stubber as stubber:
+            self.add_describe_endpoints_response(stubber, discovered_endpoint)
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(
+                os.environ.get('AWS_ENDPOINT_DISCOVERY_ENABLED'), "True"
+            )
+            self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_required_with_env_var_disabled(self):
+        self.environ['AWS_ENDPOINT_DISCOVERY_ENABLED'] = "False"
```

Comment:
For one of these true/false cases, it would be interesting if we swiched the casing to lowercase `false` or `true` to make sure those hold up.